File: src/replay_state.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from src.commentary_replay import ReplayMode, ReplayScheduleItem
from src.score_progression import ScoreProgressionEngine


@dataclass(frozen=True)
class ReplayState:
    replay_mode: ReplayMode
    elapsed_seconds: float
    match_minute: int
    current_scene: ReplayScheduleItem | None
    visible_scenes: tuple[ReplayScheduleItem, ...]
    home_score: int
    away_score: int


class ReplayStateEngine:
    """Builds the spectator-facing state at any replay timestamp."""

    def __init__(
        self,
        fixture: dict[str, Any],
        events: tuple[dict[str, Any], ...] | list[dict[str, Any]],
        schedule: tuple[ReplayScheduleItem, ...] | list[ReplayScheduleItem],
        replay_mode: ReplayMode,
    ) -> None:
        self.schedule = tuple(schedule)
        self.replay_mode = replay_mode
        self.score_engine = ScoreProgressionEngine(fixture, events)

        # The replay clock represents continuous match time.
        # Keep at least a regulation 90-minute match, while allowing
        # official stoppage-time events (e.g. 92') to extend the clock.
        scene_minutes = tuple(
            item.scene.match_minute
            for item in self.schedule
            if item.scene.match_minute >= 0
        )
        self.match_duration_minutes = max(
            90,
            max(scene_minutes, default=90),
        )

    def state_at(self, elapsed_seconds: float) -> ReplayState:
        if elapsed_seconds < 0:
            raise ValueError("Replay time cannot be negative")

        elapsed_seconds = min(
            float(elapsed_seconds),
            float(self.replay_mode.total_seconds),
        )

        visible: list[ReplayScheduleItem] = []

        for item in self.schedule:
            if item.replay_seconds <= elapsed_seconds:
                visible.append(item)
            else:
                break

        current_scene = visible[-1] if visible else None

        # Continuous match clock:
        # every real match minute has the same replay duration.
        progress = (
            elapsed_seconds / float(self.replay_mode.total_seconds)
            if self.replay_mode.total_seconds > 0
            else 0.0
        )

        match_minute = min(
            self.match_duration_minutes,
            int(progress * self.match_duration_minutes),
        )

        score = self.score_engine.score_at(match_minute)

        return ReplayState(
            replay_mode=self.replay_mode,
            elapsed_seconds=elapsed_seconds,
            match_minute=match_minute,
            current_scene=current_scene,
            visible_scenes=tuple(visible),
            home_score=score.home_score,
            away_score=score.away_score,
        )

File: web/src/lib/replay/replayState.ts
import type {
  ReplayMode,
  ReplayScheduleItem,
  ReplayState,
  ReplayTacticsState,
} from "./types";

import {
  REPLAY_MODE_SECONDS,
} from "./replaySchedule";

import {
  ScoreProgressionEngine,
  type ScoreProgressionEvent,
  type ScoreProgressionFixture,
} from "./scoreProgression";

export class ReplayStateEngine {
  private readonly schedule: ReplayScheduleItem[];
  private readonly replayMode: ReplayMode;
  private readonly scoreEngine: ScoreProgressionEngine;
  private readonly matchDurationMinutes: number;
  private readonly tactics: ReplayTacticsState;

  constructor(
    fixture: ScoreProgressionFixture,
    events: ScoreProgressionEvent[],
    schedule: ReplayScheduleItem[],
    replayMode: ReplayMode,
    tactics: ReplayTacticsState,
  ) {
    this.schedule = [...schedule];
    this.replayMode = replayMode;
    this.tactics = tactics;

    this.scoreEngine = new ScoreProgressionEngine(
      fixture,
      events,
    );

    const sceneMinutes = this.schedule
      .map((item) => item.scene.match_minute)
      .filter((minute) => minute >= 0);

    this.matchDurationMinutes = Math.max(
      90,
      ...(sceneMinutes.length > 0
        ? sceneMinutes
        : [90]),
    );
  }

  private tacticsAt(
    matchMinute: number,
  ): ReplayTacticsState {
    const resolveSide = (
      side: ReplayTacticsState["home"],
    ): ReplayTacticsState["home"] => {
      const timeline = side.timeline;

      let formation = side.formation;
      let formationId = side.formation_id;
      let playStyle = side.play_style;
      let changed = false;
      let changeMinute: number | null = null;

      for (const item of timeline) {
        const minute = Number(item.time ?? -1);

        if (!Number.isFinite(minute) || minute > matchMinute) {
          continue;
        }

        const nextFormation =
          typeof item.formation_name === "string"
            ? item.formation_name
            : formation;

        const nextFormationId =
          Number.isFinite(Number(item.formation_id))
            ? Number(item.formation_id)
            : formationId;

        const nextPlayStyle =
          typeof item.play_style === "string"
            ? item.play_style
            : playStyle;

        const formationChanged =
          nextFormation !== formation;

        const styleChanged =
          nextPlayStyle !== playStyle;

        if (formationChanged || styleChanged) {
          changed = true;
          changeMinute = minute;
        }

        formation = nextFormation;
        formationId = nextFormationId;
        playStyle = nextPlayStyle;
      }

      return {
        formation,
        formation_id: formationId,
        play_style: playStyle,
        changed,
        change_minute: changeMinute,
        timeline,
      };
    };

    return {
      home: resolveSide(this.tactics.home),
      away: resolveSide(this.tactics.away),
    };
  }

  stateAt(elapsedSeconds: number): ReplayState {
    if (elapsedSeconds < 0) {
      throw new Error(
        "Replay time cannot be negative",
      );
    }

    const totalSeconds =
      REPLAY_MODE_SECONDS[this.replayMode];

    const elapsed = Math.min(
      Number(elapsedSeconds),
      totalSeconds,
    );

    const visible: ReplayScheduleItem[] = [];

    for (const item of this.schedule) {
      if (item.replay_seconds <= elapsed) {
        visible.push(item);
      } else {
        break;
      }
    }

    const currentScene =
      visible.length > 0
        ? visible[visible.length - 1]
        : null;

    const progress =
      totalSeconds > 0
        ? elapsed / totalSeconds
        : 0;

    const matchMinute = Math.min(
      this.matchDurationMinutes,
      Math.floor(
        progress *
          this.matchDurationMinutes,
      ),
    );

    const score = this.scoreEngine.scoreAt(
      matchMinute,
    );

    return {
      replay_mode: this.replayMode,
      elapsed_seconds: elapsed,
      match_minute: matchMinute,
      current_scene: currentScene,
      visible_scenes: visible,
      home_score: score.home_score,
      away_score: score.away_score,
      tactics: this.tacticsAt(matchMinute),
    };
  }
}